### DISCUSSION TITLE
New major release and plans for 1.16.x and 1.14.x

### DISCUSSION THREAD

**Discussion by mrjimenez** [pupnp/pupnp#230] [General]
Hi Folks,

We are currently planning a new major release. The reason for a major release is that recent commits have introduced the use of a CMake build, which in turn made it possible to maintain a windows build. In order to avoid a conflict with windows libraries, we cannot use the current library name and so it has been proposed a name change to `pupnp`, which is in fact the name of the project. Personally I have no problems with that. @Vollstrecker has contacted some distro maintainers to see what would be the impact, and it seems like it would be mostly none.

However, that name change is still not in place. I have created a `branch-1.14.x` branch which I intend to be a temporary backport only branch, i.e., bug-fixes, security issues and new stuff that does not break the API are committed to `master` and back-ported to `branch-1.14.x`. Also, as you might have noticed, 1.14.1 has been released.

That situation would change when `1.16.0` is released. From then on,  `branch-1.14.x` would become ideally `frozen`, but possibly `security fix only`. Personally I would like that release-1.14.2 would be the last in the 1.14 series.

What do you think about this plan? All remarks are welcome.

Regards,
Marcelo.

**Comment by ffontaine** [comment]
If the library name changes, I assume that we'll also change the name of the pkg-config filename for consistency.
If the library name and/or the pkg-config filename changes, all users of pupnp will have to update their Makefile, configure.ac or meson.build.

For example, here is the list of the packages that will have to be updated on buildroot (I assume that debian or bigger distros have more packages):
 - igd2-for-linux: https://github.com/Orange-OpenSource/igd2-for-linux/blob/9c8b4f257538e19494e8b8565ae02786048ef3ed/linuxigd2/configure.ac#L26
 - gerbera: https://github.com/gerbera/gerbera/blob/f10c11e8fd6ac9baf6b3d72995e9c6123b7210f1/cmake/Findpupnp.cmake#L10
 - gmrender-resurrect: https://github.com/hzeller/gmrender-resurrect/blob/b2362c3b60991f0ed1e6493fce7de3ee2e74c8cb/configure.ac#L93
 - mpd: https://github.com/MusicPlayerDaemon/MPD/blob/bf270a5663c67f1ef4edc8656f428d01372df3a5/src/lib/upnp/meson.build#L1
 - ushare: https://github.com/ddugovic/uShare/blob/5f7f66cd89d5b0652c1226e65bbf9c85aeba00f2/configure#L648
 - vlc: https://github.com/videolan/vlc/blob/8f1edfab37ab2e1b54f21ef3a8f550159ac6852d/configure.ac#L4342

So, the change is "small" but it is not "none" especially as those packages will have to support both names for quite some time.

**Reply by Vollstrecker** [reply]
Sure, that's what a changelog is for, as a user of a lib you always should have at least one guy that looks for changes. For example I used the old name in amule's cmake script's and we releases the first version in 4 years just 2 hours before I found it. Admitted, we a most likely the only project that uses the *.cmake files atm, and I need to support both names, too,. But that's life.

And as all names get changed (just the filenames. It would be more correct to rename also the functions, but that would cause real issues), there's no problem that the distros ship 1.14.x and 1.16 in parallel. If we also rename the install-dir for the headers, even the *-dev packages are co-installable.

So scenarios possible are (from my view):
- Distros provide symlinks from the old to the new name for *pc and *.cmake, so every package using this stuff will get the updated info's without the need of doing anything. We don't provide them, as their maintainers know better how long they are needed
- Packagers patch the old sources to look for the new locations until upstream get's them in an update.

On a longer run, this get's adopted upstream, like any other change. You'll find in almost every project #ifdefs that react to old or new behaviour of libs. For example in v5.5 cryptopp moved the obseleted algorythms into a new namespace, so adding a second name to look for to the config-scripts (without the need of any codechanges) isn't something that's usually worth to bother about (as long as it doesn't happen in every version). Most projects take such things as an opportunity to refactor the stuff related to that libs, like I would at least recommend (and offer on request) to gerbera.

**Comment by whyman** [comment]
I think the rename doesn't achieve much other than creating work for users, both devs using the library and packagers.

The theoretically impacted users on windows are more likely to static link anyway (considering this is BSD code) so DLL naming is not an issue worth renaming the whole package for, or at least can have some workaround. It seems to me had this had been an issue up to now then some windows users would have reported it.

As a library a consideration around "We do not break user-space" is sensible.

What is the gain from all the churn created? Simply a working upnp library again :shrug: 

The package is _de facto_ called `libupnp`, so I would support the opposite and rename the repos to `libupnp` :smile:

**Reply by Vollstrecker** [reply]
> I think the rename doesn't achieve much other than creating work for users, both devs using the library and packagers.

Sure, and if we change a the name just on windows, the devs need to maintain a detection with the windows and the unix name. So work needs to be done anyways. But it should be easier for example in cmake to add find_package (pupnp CONFIG) and then only do the whole cascade afterwards if this isn't found, instead of duplicating the several find_header and find_library calls depending on the plattform. And yes, a packager always needs to check what get's installed, so if he (or you) decides to ship only the new version, and (temporary) symlink upnp.pc->pupnp.pc exactly nothing changes for any newly build app, so they only have to leave out to old upnp.pc when updating the old package.

> or at least can have some workaround.

A workaround to get different names on different plattforms, and then other workarounds in the other projects to get them unified again? In gerbera you use pupnp::pupnp which points to anything related to that, so you just would replicate lines 21-23 depending on the plattform, or you'll add pupnp somewhere in line 22, what takes more work? OK, it's doesn't seem that your stuff is designed for multiplattform, but it's just an example.

> It seems to me had this had been an issue up to now then some windows users would have reported it

You mean in the last years where it wasn't possible to build on win at all? It's not the only reason, but look at the howto of building vlc for win. Not possbile without crosscompile, mingw and msys. And yes, mingw ships only static libs, and they are not often updated, so all the neat IPv6 fixes you bring in, won't enter that. But back to topic, atm there is one vlc-plugin that utilizes upnp. If a second one comes up, they will be glad to have one dll instead of 2 statically linked binaries. Yes, this will be a long road, but I'm working on the basements for this.

> As a library a consideration around "We do not break user-space" is sensible.

For users: Right. For dev that use it: It's always better to make a clean break than trying to keep the backwards compat stuff clean an functional. That's why upnpinit2 came up. Iirc is was zlib, that just renamed their prebuild libraries that dev on win can use from zlib.dll to zlib2.dll because of incompatible changes, so yes, this can happen and devs have to deal with it.

> The package is de facto called libupnp

Because packagers decided to call it that way. What if they decide npupnp is the new thing to go for and ship that as libupnp? Doesn't happen? Install default-myql-server on a debian-system, you'll get maria-db.

**Reply by ffontaine** [reply]
Actually, the package is named libupnp because Intel decided to name it that way in 2000.
Most of the time, the package name is not defined by buildroot/debian/... packager but by the upstream project.
It should also be noted that "pupnp" is still referenced as libupnp in the NIST NVD (for security vulnerabilities): https://nvd.nist.gov/products/cpe/detail/753253?namingFormat=2.3&orderBy=CPEURI&keyword=libupnp&status=FINAL

I'm not against this change but if we go that path, I think we should at least add a symlink libupnp.pc -> pupnp.pc.

**Reply by Vollstrecker** [reply]
I'm talking about things like: https://sources.debian.org/src/pupnp-1.8/1:1.8.4-2/debian/control/#L16

And yes, this was also somewhere around the time the conflicting upnp.dll was inserted by MS. Maybe it's the same, that somehow got forked later by them, or they just ship a really old version.

And yep, that's what I meant with the symlinks. Just that (at least the 3 distro-packagers that answered) prefer to add these links by themselfs. It's just because some prefer to patch the name the other software searches for, or they provide the old and the new lib for a smooth transition. Think about you have 2 programs that use the old and 1 the new name. Usually they could start providing  upunp-1.14 and upnp-1.16 packages, that are fine to be installed side-by-side, but there is only one upnp-dev package for now. They can now decide if the keep the old one with that name and make upnp-16-dev, to be able to let the old and new ones build. Maybe they (or we) rename the include dir to include/upnp-16 in the package (possible with config-param), or the old one to include/upnp-old, or whatever they want to.

More practical they just decide, that they keep both binary-packages, and keep providing just one upnp-dev, as then no one needs to adjust their build-deps. They just will add that symlink and every package will get redirected to the new locations (except the one that decided to not use that informations and search the lib by themself by name, but then they implicit decided that they want to take care about that by themselfs, too. There is a reason for things like pkg-config). At least in debian I know it, after building the scripts decide which binary dependencie they enter in the package-description.

When the other 2 packages adopted the new names, they can just drop out the old packages and continue like nothing happend.

And the right time for doing that, is easily to find out by just getting a list of what packages list the old one in their bin-deps. If we would provide that link, we have to decide when the right time has come. I've seen many discussions in many in many projects about dropping old compat-stuff and always someone appears with "That old distribution, or that LTS release still uses it, and we want the stuff from yesterday to be useable on 6 years old systems", and then the old stuff never appears, just because noone knows all possible systems where this could still be in use. On (still debian as I know that) they just say "the next stable drops this". The few leftover packages that need it will either be patched by their packagers, or they poke upstream to finally migrate. And yes, this still happens. amule for example did the step from upnpinit to upnpinit2 just around a month ago.

After reading this again, I find it some kind of confusing. Hopefully you understand the thoughts behind this better than I was able to write them.

**Reply by mrjimenez** [reply]
> I think the rename doesn't achieve much other than creating work for users, both devs using the library and packagers.
> 
> The theoretically impacted users on windows are more likely to static link anyway (considering this is BSD code) so DLL naming is not an issue worth renaming the whole package for, or at least can have some workaround. It seems to me had this had been an issue up to now then some windows users would have reported it.
> 
> As a library a consideration around "We do not break user-space" is sensible.
> 
> What is the gain from all the churn created? Simply a working upnp library again 

All valid points to keep in mind.

> 
> The package is _de facto_ called `libupnp`, so I would support the opposite and rename the repos to `libupnp`

This is in fact a possibility. The project is called [`pupnp`](https://pupnp.sourceforge.io/) [SF project page is here](https://sourceforge.net/projects/pupnp/), because it is a fork of [`upnp`](http://upnp.sourceforge.net/), [SF project page is here](https://sourceforge.net/projects/upnp/), which is not maintained since 1.3.1, there goes a long time... Anyway, Sourceforge was almost exclusively the only sane platform to share code these days. And there could not be two projects with the same name in the whole site. This is not the case with `Github`. We can easily become `pupnp/upnp` here.

Also, the current situation is indeed messy. For some time I though OpenSUSE did not distribute `pupnp`, guess why? I was searching for `pupnp` and the project is named `upnp`. A search for `pupnp` returns `libnpupnp`. Man, this is starting to sound like a [millennium problem](https://www.claymath.org/millennium-problems/p-vs-np-problem). Just for reference, we are currently `libupnp-[devel,debugsource]` and `libupnp17`, `libupnp17-[debuginfo,32bit,32bit-debuginfo]`. Should I mention `libupnpp` and `libgupnp`? Nah, you got the point.

> It should also be noted that "pupnp" is still referenced as libupnp in the NIST NVD (for security vulnerabilities):

Which adds to this argument to rename this project to `upnp`.

On the other hand, we have the windows problem, which is the returning kid on the block. From Vollstrecker's explanation, I get that the issue is the DLL name that would require some magic on the CMake side.

Also nobody should feel intimidated by the fact that I have created a new branch. If we find it necessary, reverting `master` to `1.14.x` is a matter of changing four files since there has not been a formal 1.16 release. I did that so that we allow space for all possibilities.

Since I don't really own the project, I don't want to set a `maintainer bias`, so maybe after we have enough arguments here we can do some kind of poll within the people that has contributed to the project?

One way or the other, I am happy to see that there are people concerned with that and presenting their arguments. The last thing I want to do now with a stable library like that is to create unnecessary noise.

**Reply by Vollstrecker** [reply]
> On the other hand, we have the windows problem, which is the returning kid on the block. From Vollstrecker's explanation, I get
> that the issue is the DLL name that would require some magic on the CMake side.

On our side: Around 4 lines to have different names? On the others projects side: Unpredictable (in both cases).

> Since I don't really own the project, I don't want to set a maintainer bias, 

Who else is left over time?

**Comment by mrjimenez** [comment]
Well, we had some time to mature this idea.

The situation is the following: keeping both branches is starting to get painful, new functionality has already been added to the library and if I am getting things right, it seems that the changing the name of the library will bring more headache than profits. So my proposal is that the next release be 1.16.0 straight from `master` and that `1.14.x` is now frozen-bugfix-only.

**Reply by Vollstrecker** [reply]
I now activated Actions in my fork, so the problem is visible. https://github.com/Vollstrecker/pupnp/actions/workflows/ccpp.yml
Ian's workaround was still in there, that's why our tests worked. This workaround implies, that every prog that's using the lib either need to link statically, or to have the lib in the same folder as their binary (even a libs subdir doesn't work).

I can't see the confusions, as the 1.14 has the cmake already included, everyone out there is able to use UPNP::Shared/UPNP::Static for their linking. This is the recomended way in cmake as it pulls in the deps also (and the only way to get the build on Win anyways). What libname hides behind these aliases is of no interest to the normal programmer.

And yes, I still don't like it, but with the aliases it would be possible to have different names for different systems (compare the 4th commit in the overview)

So situation before cmake: Devs search with libnames and do some checks on the headers (on systems without pkg-config) to guess the flags.

Situation with cmake: Before the old style (if older lib is installed) they do find_package(upnp CONFIG) and then they have to decide
- if they use UPNP::Shared/UPNP::Static for linking and (maybe) create aliases with the old version to fit the new ones
- if they build the aliases from the values their old search-method gives them
- if they just drop support for the old ones at all.

Situation with different libname:
- if they've choosen option 1 before - Done

For sure this doesn't contain anything from packagers view, but most of them just don't care about such a thing if they know about it and it doesn't happen every 2 weeks.

The confusion will be bigger when people get different results depending where they install their libs.
Not to mention that I think about adding some stuff for cpack, so github could create an installer, so people that want to just compile their app, can just install the deps without comiling them by themself, which then could just contain the static lib, as the shared version just can't be used.

**Comment by mrjimenez** [comment]
Hi folks,

I made a mess for myself with those two branches, so I will try to remedy the situation.

Please correct me if I am wrong on this statement: `the only thing currently "broken" is the windows dynamic linkage that needs exception rules in CMake`.

If the above is true, I think we may proceed discussing normally as if I had not created a new branch. Next release will be 1.16.0 and life goes on on `master`, as well as this discussion.

When we reach a consensus about what to do with the naming, we implement it, be it a full rename of the lib or a partial rename on windows dynamic link. I don't consider the naming matter over, I just don't want it to be an obstacle.

Regards,
Marcelo.

**Reply by Vollstrecker** [reply]
> Please correct me if I am wrong on this statement: the only thing currently "broken" is the windows dynamic linkage

Exactly. A usual VS-project builds everything in one dir (like Ians "fix" simulates), so the tests (if they were implemented, didn't check) ran fine. And if the usual way was static-linking, or installing the progs and libs into one dir, everything is fine, too. The whole problem is a duplicated libname in combinations with a fixed search-order for libs.

> that needs exception rules in CMake.

Nope, with CMake it just got visible. The rules only affect our build, builds of other projects are not "fixed" with that.

> When we reach a consensus about what to do with the naming

I'm open for every solution, either a complete rename, or a win-only rename. I consider myself not long or deep enough member to take part in that decision, I just put the finger on the problem to point out what needs to be fixed.

> I just don't want it to be an obstacle.

Depends on how you define obstacle. If you're fine with ocassional reports about "broken" libupnp from other projects, just to find out that they installed dynamic libs in a subdir, then all is fine as it is.

**Reply by mrjimenez** [reply]
> I'm open for every solution, either a complete rename, or a win-only rename. 

Ok.

> I consider myself not long or deep enough member to take part in that decision, I just put the finger on the problem to point out what needs to be fixed.

This is not quite correct. If we have CMake in the project today it is mostly because you did it, so your opinion counts as much as any other member of the project. Antiquity is not a hierarchy here.

> Depends on how you define obstacle. If you're fine with ocassional reports about "broken" libupnp from other projects, just to find out that they installed dynamic libs in a subdir, then all is fine as it is.

No, I was not referring to that. I meant I don't want the naming problem to be in the way of new releases. I want to address it without interrupting the in-flow of code, just that.

On the matter that you addressed, what exactly happens in a windows setup when you try to run an application and it dynamically links to the wrong libupnp? Segfault or the program simply does not load? Is CMake able to detect the situation that the user would link to the wrong library and give the user an error abort during the build?

Maybe while we sort this out we should leave it documented in the home page that the library does not yet support dynamic linking on windows.

**Comment by mrjimenez** [comment]
A folow up on the new release.

1. Library name problem
2. Library handle feature
3. IXML bug
4. Remove ithread and leave it straight pthread

Lets see what we have for each of these topics:

1. Has been discussed so far, but as far as I can see, no conclusion.
2. This has been asked for in #241 and more or less implemented the basis for it in #255 . We need to decide if this is useful so that we commit it or not.
3. That is in my opinion the big deal. I have not tried to debug it yet, but a few years ago I tried to understand the IXML code and make it a little better, but I felt it needed a rewrite and abandoned the project. There is a possibility to use something working, tested and small, [Tiny XML2](https://github.com/leethomason/tinyxml2), but it is C++. On that matter, we now have #260 , which showed as a very easy path to start using mixed C/C++ in the library, so that is another big decision to make: should we make libupnp mixed code? My big concern used to be code size, but nowadays that seems no longer be a true problem.
4. The ithread code is also a legacy from a time where there was no POSIX threads standard. We have long removed the thread code from the library, but left those defines so that embedded folks could in principle plug in some other thread library. But nowadays it only complicates the life of the IDEs and our lives as a consequence. I vote for removal.

Please comment.

**Reply by whyman** [reply]
1. Stick with libupnp and rename the repo.
2. Amazing work, would be great to have it merged.
3. IXML structures are in the public interface so would be tough to remove. Maybe combine with 2 as a 2.0? 
4. There seems to be some windows pthread compat in there too... If we go with c++ we could use std::thread

**Reply by mrjimenez** [reply]
1. Lol, I did not even consider that, but I must say I like it. :laughing: 
2. Ok, I think it would be ok to merge.
3. Very good point. There is quite some work here. We should hide the data structures and keep the interface. I see that in `upnp.h` there are only `IXML_Document` and `DOMString`. The latter can be easily substituted by a better string type, maybe `UpnpString` or something else more standard if we go to C++. The former we can choose to leave an open interface to TinyXML2 or we hide it behind an opaque object and expose the minimum interface we need. If I am not mistaken, all we need is an interface which will return an XML document from an URL, please DO correct me on that if I am wrong.
4. I was not aware of std::thread, that leaves us within a standard an is one less dependency. :+1:

**Reply by mrjimenez** [reply]
I just found that TinyXML2 does not support namespaces. That's too bad, I was halfway through... :disappointed: 

TinyXML2 was really good.

Any ideas?

**Reply by whyman** [reply]
In Gerbera we are using https://pugixml.org/

**Reply by Vollstrecker** [reply]
Beside the use of PROJECT_SOURCE_DIR, they do everything I would expect for modern cmake. Inclusion for downloading can be done within a minute. So it seems it's worth a try.

**Reply by mrjimenez** [reply]
From the [pugi manual](https://pugixml.org/docs/manual.html):
```text
8.6 Conformance to W3C specification
...
. Namespace nodes are not supported (affects namespace:: axis).
. Name tests are performed on QNames in XML document instead of expanded names; 
for <foo xmlns:ns1='uri' xmlns:ns2='uri'><ns1:child/><ns2:child/></foo>, 
query foo/ns1:* will return only the first child, not both of them.
Compliant XPath implementations can return both nodes if the user provides
appropriate namespace declarations.
```

Can we get away with that? I don't remember, but I don't think IXML actually evaluates the namespaces.

**Reply by Vollstrecker** [reply]
If namespaces are no requirement, you now can choose between tiny and pugi.

**Reply by mrjimenez** [reply]
Unfortunately, they are a requirement. And I just checked, IXML does the proper namespace processing.

**Reply by ingo-h** [reply]
1. Library name problem  
With the note of @Vollstrecker, is it still an issue?
> I've tested now with a virtualbox image provided for testing by MS itself. It worked and didn't have the conflicting file. https://github.com/pupnp/pupnp/discussions/224#discussioncomment-369696

2. Library handle feature  
I have made a suggestion https://github.com/pupnp/pupnp/discussions/267

3. IXML bug  
Are there some details of the problem?

4. Remove ithread and leave it straight pthread  
I would say to clean up the code and remove ithread

**Reply by mrjimenez** [reply]
1. No longer an issue.
3. I am taking a look right now, I will answer there.
3. #249
4. :+1:

**Reply by ePirat** [reply]
Regarding 3. the only really mature and complete open source XML library I am aware of is libxml2, but due to how libxml2 initialization/teardown works, it is incredibly challenging to properly use in a library, if not impossible even.

**Reply by mrjimenez** [reply]
Item 3 has hopefully been fixed in #306 . Maybe we should stick to ixml and improve it in the weak areas.

**Comment by mrjimenez** [comment]
Hi Folks, hope you are all doing well.

There is an issue I wanted to address since we dealt with the Callstranger security advisor. We should deal with it because since master has been reformated with spaces instead of tabs, backporting has been a nighmare for me. BTW, I'd rather use tabs for C/C++, but I digress.

But first I must introduce the problem, with a story to justify my error.

I have started coding a long time ago when windows dynamic libraries were 16 bit stuff. And some conceptions became misconceptions when I started using linux. To make a long story short, I think that the library handle thing introduced in master is most likely a mistake. There is a confusion about handles because these handles are process wise, not system wise. In my mind, the dynamic library data segment was shared among processes. This is not the case, not in unix dynamic library and not in windows 32 dynamic libraries. So the coding is likely a bad mistake that should be reviewed and a decision should be made.

The decision is: fix master or abandon it and continue to 1.18 from 1.14 and eventually code it properly from there? I have been trying to backport everything to 1.14.x, but it does not make sense to carry new features to that branch and it does not make sense to continue on master if we decide to step back on the library handle thing.

Regards,
Marcelo.

**Reply by whyman** [reply]
The first thing I want to say is that its up to you Marcelo.

My understanding for the motivation was to _abstract and encapsulate the global state_ that the library maintains and to not pollute it into the users application - and from that perspective I think it makes sense and is a successful change.

Having an API break is not fun though, and the further the branches diverge the harder backporting will become as you have seen.

As for the the tabs, we can and should change that easily thanks to clang-format.

**Reply by mrjimenez** [reply]
I know the motivation was ok, but I might have hugely overdone it and caused a major API breakage in the process.

All the other things we did were ok and I think we should continue the development, but really, the API breakage is worrying, especially if it is not needed.

**Reply by whyman** [reply]
If you have some idea that still allow hitting those goals without the API break I am keen to hear it - I am not sure it would be possible.

**Reply by whyman** [reply]
So whats the plan then? Rename current master to archived-master and then push master from the 1.14x branch and continue development there?

**Reply by mrjimenez** [reply]
That seems to be the best course, but only if you agree with my concern.

I don't know if I made it clear what exactly is the problem. I think that all the work done in master has been great and I don't want to loose it. The only exception is this library handle thing, which has an immediate API breakage consequence and is likely overdone, but I don't have enough spare time right now to deal with it. And postponing this analysis is only making the problem worse.

I don't know what you think guys, that is what I would like to hear.

> If you have some idea that still allow hitting those goals without the API break I am keen to hear it - I am not sure it would be possible.

Maybe we can open a thread to discuss this particular matter. I am having some difficulties in findind the thread that originated this idea. Pull request #255 has to do with it, but I remember that the original discussion was about something more specific.

**Reply by whyman** [reply]
Fine with me, backports shouldn't be too hard

**Comment by Vollstrecker** [comment]
From my side there shouldn't be any problem, I only remember the splitting into smaller file in cmake/ which should be easy portable (I've no clue why I didn't consider to backport this). Maybe the functions for the checks had been reworked, but I think this shouldn't matter for now. Do the gtest tests also work without the handle, or are they lot for now?

**Reply by mrjimenez** [reply]
> From my side there shouldn't be any problem, I only remember the splitting into smaller file in cmake/ which should be easy portable (I've no clue why I didn't consider to backport this). 

Ok.

> Maybe the functions for the checks had been reworked, but I think this shouldn't matter for now. Do the gtest tests also work without the handle, or are they lot for now?

Should work without the handle.
